processing/src/main/java/org/apache/druid/segment/serde/ColumnPartSerde.java

```diff
@@ -39,13 +41,23 @@
     @JsonSubTypes.Type(name = "floatV2", value = FloatNumericColumnPartSerdeV2.class),
     @JsonSubTypes.Type(name = "longV2", value = LongNumericColumnPartSerdeV2.class),
     @JsonSubTypes.Type(name = "doubleV2", value = DoubleNumericColumnPartSerdeV2.class),
+    @JsonSubTypes.Type(name = "null", value = NullColumnPartSerde.class)
 })
 public interface ColumnPartSerde
 {
   @Nullable
   Serializer getSerializer();
 
-  Deserializer getDeserializer();
+  /**
+   * Returns a Deserializer to read a column from a segment.
+   *
+   * @param rowCountSupplier          a supplier that returns the row count of the segment that the column belongs to.
+   *                                  Getting from the supplier the row count can be expensive and thus should be
+   *                                  evaluated lazily.
+   * @param segmentBitmapSerdeFactory bitmapSerdeFactory stored in the segment. Each columnPartSerde may have their own
+   *                                  bitmapSerdeFactory.
+   */
+  Deserializer getDeserializer(IntSupplier rowCountSupplier, BitmapSerdeFactory segmentBitmapSerdeFactory);
```

Comment:
hmm, this seems a bit strange of a change

In most implementations, i think these things are basically part of the `ColumnPartSerde` when it is created (and the bitmap serde factory is shared across all columns via the index spec), why is it passed into creating the deserializer?

Reply:
> these things are basically part of the `ColumnPartSerde` when it is created (and the bitmap serde factory is shared across all columns via the index spec)

Actually, this is the reason that I pass `segmentBitmapSerdeFactory` in. Even though the bitmapSerdeFacotry is shared across all columns, each column stores it again in its columnPartSerde which is duplicate with the bitmapSerdeFactory stored in the segment. This seems unnecessary and I wanted to avoid it for `NullColumnPartSerde`.

Reply:
I think I'd side on the thinking that this PR shouldn't be changing this interface.  It is perhaps sub-optimal that we are relying on Jackson to inject in the factory which means that we likely are storing a bunch of extra references (and maybe even allocating a bunch of extra objects) than we really should.  This interface change as done in this PR is trying to make life "better" for the null column, but doesn't actually fix it for all of the other columns.  Even when we do that fix, I'd still advocate for us figuring out a way to fix it through the constructors of the ColumnPartSerde objects instead of by changing the interface.

As it stands, knowing the number of rows is definitely something that should be done via the constructor.  When this is persisted, it knows the number of rows, it can store that as part of the constructor and doesn't need an extra parameter on the interface method.  

For the BitmapSerdeFactory, an argument could be made for things to go in various different ways.  I *think* that we might be getting to the point where we actually want a Factory involved in here that can be given things like metadata about the segment as a whole.  But, that's just random musings.  I'd suggest that we keep to the current model for now and queue up a change to how these factories are dealt with at a later stage.

Reply:
So, here are the reasons that I modified this interface.

- I added `NullColumnPartSerde` as a default type of `ColumnPartSerde`. This will be useful when the historical tries to deserialize a `ColumnPartSerde` that is unknown to the version of the historical. The historical will use `NullColumnPartSerde` instead of exploding.
- Since `NullColumnPartSerde` is the default type of `ColumnPartSerde`, its constructor should be simple with no parameters. The fallback to `NullColumnPartSerde` for unknown `ColumnPartSerde` will fail otherwise.
- To create a bitmapIndex for null columns, `NullColumnPartSerde` needs to know the number of rows in the segment and the bitmap index type. 

I agree that the current interface change is specific to this PR which could be better if it is more generic. I have reverted this particular interface change in this PR but am going to make another PR for it. In this PR, [I made the null column _not-filterable_](https://github.com/apache/druid/pull/12250/files#diff-753cc3c64ec70b8ca1be35049eb1273be6403fd3ab447ba6bdc0e234c75646b1R91-R94), so that `ColumnSelectorBitmapIndexSelector` can create one for it instead. This is a hacky workaround to prioritize fixing `getDeserializer` interface.